### PR TITLE
Fix memory leak in metrics chain element

### PR DESCRIPTION
**Pull request by Ex4amp1e**
## Description
Temp data has not been cleaning up in the previousMetrics map after connection close.

## Issue link
https://github.com/networkservicemesh/sdk/issues/1620

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [x] Bug fix
- [ ] New functionality
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/metrics/options.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2022-2024 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/metrics/options.go:1]:
```suggestion
// Copyright (c) 2024 Cisco and/or its affiliates.
```

File: pkg/networkservice/common/metrics/server.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
-// Copyright (c) 2021-2022 Doc.ai and/or its affiliates.
-// Copyright (c) 2023 Nordix Foundation.
+// Copyright (c) 2021-2024 Doc.ai and/or its affiliates.
+// Copyright (c) 2023-2024 Nordix Foundation.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/metrics/server.go:2]: Do not edit exits copyrights

**Review by denis-tingaikin — COMMENTED** [review]
TODO:

1. try to use metadata for storing preivous metric values
2. add chnages https://github.com/networkservicemesh/sdk/pull/1625

File: pkg/networkservice/common/metrics/server.go (hunk 2 above)
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/server.go]: Fixed

File: pkg/networkservice/common/metrics/options.go (hunk 1 above)
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/options.go]: Done

**Review by denis-tingaikin — COMMENTED** [review]
Added a few comments; the changes mostly look good.

File: pkg/networkservice/common/metrics/metadata.go [hunk 3]
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) 2022 Cisco and/or its affiliates.
 // Copyright (c) 2023 Nordix Foundation.
+// Copyright (c) 2024 Cisco and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/metadata.go]: Done

File: pkg/networkservice/common/metrics/metadata.go [hunk 4]
```diff
@@ -19,16 +20,21 @@ package metrics
...
 	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/metadata"
 )
 
 type keyType struct{}
-type metricsMap = map[string]metric.Int64Counter
 
-func loadOrStore(ctx context.Context, metrics metricsMap) (value metricsMap, ok bool) {
+type dataType struct {
+	counter  map[string]metric.Int64Counter
+	previous *sync.Map
```
[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/metadata.go]: Done

File: pkg/networkservice/common/metrics/metadata.go [hunk 5]
```diff
@@ -19,16 +20,21 @@ package metrics
...
 	"go.opentelemetry.io/otel/metric"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/metadata"
 )
 
 type keyType struct{}
-type metricsMap = map[string]metric.Int64Counter
 
-func loadOrStore(ctx context.Context, metrics metricsMap) (value metricsMap, ok bool) {
+type dataType struct {
```
[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/metadata.go]: Done

[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/metadata.go]: Done

File: pkg/networkservice/common/metrics/metadata.go [hunk 6]
```diff
@@ -19,16 +20,21 @@ package metrics
...
+	counter  map[string]metric.Int64Counter
+	previous *sync.Map
+}
+
+func loadOrStore(ctx context.Context, metrics dataType) (value dataType, ok bool) {
```
[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/metadata.go]: Done

File: pkg/networkservice/common/metrics/server.go [hunk 7]
```diff
@@ -72,7 +79,11 @@ func (t *metricServer) writeMetrics(ctx context.Context, path *networkservice.Pa
 				continue
 			}
 
-			metrics, _ := loadOrStore(ctx, make(map[string]metric.Int64Counter))
+			k := dataType{
+				counter:  make(map[string]metric.Int64Counter),
+				previous: new(sync.Map),
```
[the remark this answers is not in the document]
Comment (Ex4amp1e) [reply pkg/networkservice/common/metrics/server.go]: Done

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**